### PR TITLE
Sprout -> Sapling migration status nullptr defence

### PR DESCRIPTION
**Pull request by Eirik0**

### CODE REVIEW

**Comment by Eirik0** [comment]
It was discovered that when calling `z_getmigrationstatus` while migration transactions were in the mempool would cause nullptr to be dereferenced. The issue was was that the transactions are not yet associated with blocks, ad we were trying to get block times for nonexistent blocks. In fixing this issue, it was also discovered that the `unmigrated_amount` was incorrect in this situation - this is also fixed in the first commit of this PR.

**Comment by Eirik0** [comment]
The second commit in the PR is defensive. We were concerned that storing the sapling migration async operation object could run into a race condition when trying to cancel it. The fix for this was to sore the async operation's id rather than the operation itself, and then to go through the async rpc queue to recall, cancel, and remove migration operations which should not continue to execute.

As a note -
There is an independent issue: It is not possible to cancel async operations while they are running. `AsyncRPCOperation` provides a mechanism for creating custom cancel methods - one is to overrides the cancel method. The way the interface is currently designed, this is required to create operations that are cancellable while they are running. This issue is the cancel method is not declared as virtual, so this does not actually work, and, for whatever reason, adding the virtual keyword to this method exposes a circular dependency and causes the build to break. This needs to be fixed, but because of potential complexity. it should be done as a separate issue.

**Comment by Eirik0** [comment]
Created https://github.com/zcash/zcash/issues/3988.

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -3967,7 +3967,8 @@ UniValue z_getmigrationstatus(const UniValue& params, bool fHelp) {
     {
         std::vector<CSproutNotePlaintextEntry> sproutEntries;
         std::vector<SaplingNoteEntry> saplingEntries;
-        pwalletMain->GetFilteredNotes(sproutEntries, saplingEntries, "", 1);
+        std::set<PaymentAddress> zaddrs;
```
Comment (daira) [inline src/wallet/rpcwallet.cpp]: I suggest renaming this to `noFilter`.

Comment (daira) [reply src/wallet/rpcwallet.cpp]: Also add a comment saying what this is intending to do. If I'm not mistaken, it's intended to find unspent notes sent in transactions at any chain depth or in the mempool, to any address for which we have a spending key, including locked notes.

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -4000,7 +4001,6 @@ UniValue z_getmigrationstatus(const UniValue& params, bool fHelp) {
                 continue;
             }
             migrationTxids.push_back(txPair.first.ToString());
-            CBlockIndex* blockIndex = mapBlockIndex[tx.hashBlock];
             //  A transaction is "finalized" iff it has at least 10 confirmations.
             // TODO: subject to change, if the recommended number of confirmations changes.
             if (tx.GetDepthInMainChain() >= 10) {
```
Comment (daira) [inline src/wallet/rpcwallet.cpp]: Is there a constant we can use instead of hardcoding `10` here?

Comment (Eirik0) [reply src/wallet/rpcwallet.cpp]: I was not able to find a constant for this.

**Comment by Eirik0** [comment]
@zkbot r+

**Pull request merged**